### PR TITLE
Use Pydantic to validate and manage settings

### PR DESCRIPTION
**Pull request by tiran**
Pydantic provides a convenient way to validate data and transform it into type-safe Python objects with defaults.

### CODE REVIEW

File: src/fromager/settings_model.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+from . import overrides
+
+# build variant
+Variant = typing.NewType("Variant", str)
+
+# Package validates and transforms name to canonicalized name
+Package = typing.Annotated[
+    NormalizedName,
+    BeforeValidator(lambda pkg: canonicalize_name(pkg, validate=True)),
```
Comment (dhellmann) [inline src/fromager/settings_model.py:22]: To be backwards compatible, we need to use `overrides.pkgname_to_override_module()`.

File: src/fromager/settings_model.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    pre_built: dict[Variant, PackageSet] = Field(default_factory=dict)
+    """Mark packages as pre-built"""
+
+    @field_validator("packages", "pre_built", mode="before")
+    @classmethod
+    def before_none_to_default(cls, v: typing.Any, info: ValidationInfo) -> typing.Any:
+        """Pre-hook to map None to default factory result"""
+        return _map_none_to_default(cls, v, info)
+
+    def get_package(
```
Comment (dhellmann) [inline src/fromager/settings_model.py:115]: Maybe this should be "get_build_info"?

**Review by dhellmann — COMMENTED** [review]
I like this direction! I had a couple of small comments, but I'll wait for a more complete review until you remove the draft flag.

File: src/fromager/settings_model.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+PackageSet = typing.Annotated[
+    set[Package],
+    BeforeValidator(lambda v: v if v is not None else set()),
+]
+
+# URL with templating
+UrlTemplate = typing.NewType("UrlTemplate", HttpUrl)
+
+
+def _map_none_to_default(cls, v: typing.Any, info: ValidationInfo) -> typing.Any:
```
Comment (shubhbapna) [inline src/fromager/settings_model.py:35]: instead of `typing.Any` could we use a generic type?

**Review by shubhbapna — COMMENTED** [review]
Looks great! we can try porting the remaining methods and trigger the tests?

**Pull request closed without merging**